### DISCUSSION TITLE
Adaptation Keiser M3

### DISCUSSION THREAD

**Discussion by serbasan** [doudar/SmartSpin2k#494] [Q&A] [answered]
I was thinking about controlling my Keiser M3 with some buttons and a stepper motor and I found this project.
You could go so far as to modify the housing to house all the components and have the motor wind a steel cable and pull the magnet assembly. In [this video](https://www.youtube.com/watch?v=Fnh547nw1nk)  you can see how it moves with the hand.

Regarding the software part. Since the movement path would be very short, is this current firmware suitable for such a short movement? Could it be adjusted to make such small steps to adjust the resistance?

I'm waiting for help to know if I could start making a hardware adaptation.

Thank you.

**Comment by doudar** [comment]
> I was thinking about controlling my Keiser M3 with some buttons and a stepper motor and I found this project.
> 
> You could go so far as to modify the housing to house all the components and have the motor wind a steel cable and pull the magnet assembly. In [this video](https://www.youtube.com/watch?v=Fnh547nw1nk)  you can see how it moves with the hand.
> 
> 
> 
> Regarding the software part. Since the movement path would be very short, is this current firmware suitable for such a short movement? Could it be adjusted to make such small steps to adjust the resistance?
> 
> 
> 
> I'm waiting for help to know if I could start making a hardware adaptation.
> 
> 
> 
> Thank you.

Absolutely! There's a user parameter incline multiplier as well as shift step. Between those two, you can use very small (or large) movements if needed. 

Sounds like a very neat idea! The Keiser bike is popular and many people would probably like to have some sort of hardware solution for it!

**Reply by serbasan** [reply]
Another query. In my Keizer M3 I already have power and cadence via Bluetooth through another board in [this project](https://github.com/ajs123/KBikeBLE/). This works very well and is very well calibrated. I wouldn't like to lose it.

Would it work with SmartSpin2k by reading the power and cadence data that I have now or is new power and cadence data supplied by the SmartSpin2k board generated?

It hasn't been very clear to me how it works.

Thank you.

**Reply by serbasan** [reply]
I answer myself. I already found out how she matches. SmartSpin2k will act as a Bluetooth multiplexer. Brilliant!
I'll try it with my setup, hope it captures the cadence and power.

**Comment by serbasan** [comment]
I have made a number of modifications to the hardware to adapt it to the Keizer M3. I have also created an adapter that easily screws onto the rear magnet mount so I can transfer the motor spin from the smartspin2k.

The belt that transmits the rotation does not need to be closed creating tension, just one side of the belt is enough and we avoid creating tension in the whole. This is possible because the magnet has a strong spring that places the magnet in its rest position.

After adjusting the values in the settings I have achieved a very good and pleasant configuration to train in glory. Many thanks to the creator, this is wonderful.

I remember that I replaced the original screen with another great project for the Keizer M3 that adds Bluetooth and greatly improves the power measurement. https://github.com/ajs123/KBikeBLE.

These two projects together are a perfect complement to the Keizer.

I would only need the firmware to allow two things.

1. “Sim Mode Shift Amount” its minimum value of “50” is very high for Keizer since one press changes two three gears.

2. Be able to keep the engine on when connecting smartspin2k. This would allow you to use the buttons without having to use an application. As the magnet has a spring when you go up a gear, the motor is disconnected and the spring returns you to gear 0

I would appreciate help on how to modify these values in the firmware. Is it a better place to comment on “Issues” in case it could be added to the default firmware?

Thank you very much for this great project.


![2023-12-11 23 46 39](https://github.com/doudar/SmartSpin2k/assets/21089192/5dfd5fb6-8671-4d38-a0b6-6f99c701ddb1)
![2023-12-11 23 46 45](https://github.com/doudar/SmartSpin2k/assets/21089192/d10b52f8-07f8-4d8f-a5de-14fc684fe9fb)
![2023-12-11 23 46 50](https://github.com/doudar/SmartSpin2k/assets/21089192/a28ba9bf-f362-48ef-b792-17172e7e49d1)
![2023-12-11 23 46 55](https://github.com/doudar/SmartSpin2k/assets/21089192/8c87d4ce-7b55-4119-8566-87f01eeb6052)
![2023-12-11 23 47 01](https://github.com/doudar/SmartSpin2k/assets/21089192/1945f166-205d-481b-b23b-1da5cb1bd5ce)
![2023-12-11 23 47 06](https://github.com/doudar/SmartSpin2k/assets/21089192/86336227-9ce2-4a2a-9fa9-e6ff726a4780)

**Reply by doudar** [reply]
Nice writeup and pictures! What I would do to change shift steps, since your value is already very small, instead of decreasing it further, I'd increase [the micro stepping in main](https://github.com/doudar/SmartSpin2k/blob/5115f7c7febc5b51d52ea278428e5e80f5164a84/src/Main.cpp#L438-L439). Valid values should be 2, 4, 8, 16, 32, 64, 128, or 256. Each doubles the steps required for a full revolution of the stepper motor from the previous setting.

If you're on Facebook, any chance you can post this writeup on the SmartSpin2k group as well? If not, can I post it?  Fantastic job!

**Reply by serbasan** [reply]
Thank you. I'm going to try to compile by changing that value. For now I have downloaded the project and when trying to compile it I get this error. Any clues as to what may be happening?


`Processing release (platform: espressif32 @ 6.0.1; board: esp32doit-devkit-v1; framework: arduino)
-------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32doit-devkit-v1.html
PLATFORM: Espressif 32 (6.0.1) > DOIT ESP32 DEVKIT V1
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 @ 3.20006.221224 (2.0.6) 
 - tool-esptoolpy @ 1.40400.0 (4.4.0) 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Traceback (most recent call last):
  File "C:\Users\lliurex\Downloads\SmartSpin2k-develop (1)\SmartSpin2k-develop\git_tag_macro.py", line 3, in <module>
    subprocess.run(["git", "fetch", "--unshallow"])
  File "C:\Users\lliurex\.platformio\python3\Lib\subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\lliurex\.platformio\python3\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Users\lliurex\.platformio\python3\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [WinError 2] El sistema no puede encontrar el archivo especificado
OSError: 'python git_tag_macro.py' exited 1:
  File "C:\Users\lliurex\.platformio\penv\Lib\site-packages\platformio\builder\main.py", line 173:
    env.SConscript("$BUILD_SCRIPT")
  File "C:\Users\lliurex\.platformio\packages\tool-scons\scons-local-4.5.2\SCons\Script\SConscript.py", line 598:
    return _SConscript(self.fs, *files, **subst_kw)
  File "C:\Users\lliurex\.platformio\packages\tool-scons\scons-local-4.5.2\SCons\Script\SConscript.py", line 285:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "C:\Users\lliurex\.platformio\platforms\espressif32\builder\main.py", line 312:
    target_elf = env.BuildProgram()
  File "C:\Users\lliurex\.platformio\packages\tool-scons\scons-local-4.5.2\SCons\Util\envs.py", line 242:    
    return self.method(*nargs, **kwargs)
  File "C:\Users\lliurex\.platformio\penv\Lib\site-packages\platformio\builder\tools\piobuild.py", line 61:  
    env.ProcessProgramDeps()
  File "C:\Users\lliurex\.platformio\packages\tool-scons\scons-local-4.5.2\SCons\Util\envs.py", line 242:    
    return self.method(*nargs, **kwargs)
  File "C:\Users\lliurex\.platformio\penv\Lib\site-packages\platformio\builder\tools\piobuild.py", line 118: 
    env.ProcessFlags(env.get("BUILD_FLAGS"))
  File "C:\Users\lliurex\.platformio\packages\tool-scons\scons-local-4.5.2\SCons\Util\envs.py", line 242:
    return self.method(*nargs, **kwargs)
  File "C:\Users\lliurex\.platformio\penv\Lib\site-packages\platformio\builder\tools\piobuild.py", line 228: 
    env.Append(**env.ParseFlagsExtended(flags))
  File "C:\Users\lliurex\.platformio\packages\tool-scons\scons-local-4.5.2\SCons\Util\envs.py", line 242:    
    return self.method(*nargs, **kwargs)
  File "C:\Users\lliurex\.platformio\penv\Lib\site-packages\platformio\builder\tools\piobuild.py", line 190: 
    for key, value in env.ParseFlags(str(raw)).items():
  File "C:\Users\lliurex\.platformio\packages\tool-scons\scons-local-4.5.2\SCons\Environment.py", line 1035: 
    do_parse(arg)
  File "C:\Users\lliurex\.platformio\packages\tool-scons\scons-local-4.5.2\SCons\Environment.py", line 887:  
    arg = self.backtick(arg[1:])
  File "C:\Users\lliurex\.platformio\packages\tool-scons\scons-local-4.5.2\SCons\Environment.py", line 798:  
    raise OSError("'%s' exited %d" % (command, status))
======================================== [FAILED] Took 1.25 seconds ========================================

Environment    Status    Duration
-------------  --------  ------------
release        FAILED    00:00:01.254
=================================== 1 failed, 0 succeeded in 00:00:01.254 ===================================

 *  The terminal process "C:\Users\lliurex\.platformio\penv\Scripts\platformio.exe 'run'" terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it. `

**Reply by doudar** [reply]
Did you install the PlatformIO extension? @eMadman has rebuilt his VSCODE environment lately, possibly he has some input?

**Reply by doudar** [reply]
Also, here is the code block where you can prevent the roll back. Just set them both to false. https://github.com/doudar/SmartSpin2k/blob/5115f7c7febc5b51d52ea278428e5e80f5164a84/src/Main.cpp#L364-L370

**Comment by eMadman** [comment]
> Did you install the PlatformIO extension? @eMadman has rebuilt his VSCODE environment lately, possibly he has some input?

You'll also want C/C++ and cmake extensions.  I can't remember if they're all needed but this combination works for me.  
![image](https://github.com/doudar/SmartSpin2k/assets/24726844/8ccdb6fe-ed02-4b1b-bd08-4c8e41f3921b)

**Reply by serbasan** [reply, marked as answer]
With GitHub Desktop I was able to download the project and from there open VScode and it compiled.
Downloading the project folder and opening it from VSCode was when it gave those errors.
I start trying to change those values and do tests.
I'll comment on feelings.
Thank you.

**Comment by serbasan** [comment]
I'm doing tests and with 8 motorcycle steps, each press changes one gear up to the 24 there are. Brilliant.

Now I'm making my training plans and I have a problem. When I go from, for example, 195W to 85W, the gears drop to 1 and I lose many watts quickly, the ERG is deactivated and it is very difficult to activate the ERG again. It's like falling into a pit with no way out. I try to use the up gear button, I manage to activate the ERG and just at that moment it suddenly drops to 1 and I fall into the pit again.

If the workouts are more stable 180W to 130W with cadence changes everything seems to work well.

I have played with the sensitivity of the ERG and it seems that with very low sensitivities it works better (1.7) but it is not enough.

I think it is due to the short travel of the magnet and the fact that there is also a spring that pushes gear 1.

Is there a possibility that ERG is very slow when the watts drop? Somewhere in the code that I can modify some value and test.

When ERG goes up in watts there doesn't seem to be any problem.

Thank you very much for your time.

**Reply by doudar** [reply]
I'd give the Virtual Shifting branch a try and see if it preforms better for you. It has some significant improvements in how big target changes are handled in ERG mode. It will hopefully be merged into develop in the next few days. https://github.com/doudar/SmartSpin2k/tree/VirtualShiftingPassthrough

Also, you could try setting min brake watts (on the settings web page) to a higher value. That's supposed to keep your resistance from going too low.

**Reply by serbasan** [reply]
Min Bike Brake Watts seems to have done the magic. She had it at about 50w and decided to raise it to 65w and it seems that the watt drop has disappeared and the ERG does not deactivate.

I am very excited about this project, a shame that the viruses at my daughter's school don't let me train :)

I'm going to try that other branch that you told me about. What does VirtualShiftingPassthrough mean? Does this add virtual gear shifting like the Zwift Hub One trainer? I have also seen that a torque table is now created. How does that change from what we have in the main branch?

Thank you.

**Reply by doudar** [reply]
The torque table is a persistent table (boot to boot) table that stores automatically calculated power curve information about your trainer. 

The Virtual shifting branch lets you connect any FTMS trainer and use virtual shifting with it. So, if you have a regular smart trainer, you could connect our shifters to an esp32, then (via BLE) the trainer to the esp32 and it will “shift” the trainer without having to mechanically shift the bike.

**Comment by MortenVinding** [comment]
Wow this is very interesting!

I'm using my own fork of BLEBike:
https://github.com/MortenVinding/BLEBike

Very similar to KBikeBLE I believe, only it uses a ESP32 (which I already had handy).

I have a Proteus Daytona V6 which already has motorised resistance control, but only very simple with a few build in programs.
So I'm looking for a way to control resistance and also have the cadence/power in Zwift.

Did you ever find a way to connect the cadence sensor to SmartSpin2k? I would really like to avoid the expenses of a powermeter, especially as my bike already report power (based on resistance and cadence RPM of course).

Given that SS2k already support Peloton's powermeter, I can't imagine it should be that difficult to make a power calculation based on a cadence input on a GPIO pin.

My bike also has build-in heart rate sensor, do you know if that is possible to connect also (I already have that in BLEBike also)?

**Comment by serbasan** [comment]
Hello. The Keiser M3 already has a built-in sensor that measures cadence and also has a sensor that measures the position of the magnet. With this data it already estimates the power. All of this is already included in the Keiser M3 and is displayed on its screen in addition to calories, kilometers, heart rate, etc. The KBikeBLE project replaces the original screen and connects the cable that carries this data to the Adafruit nrf52840. The project firmware shows the power and cadence (only this data) in a better calculated way and transmits it via Bluetooth.
SmartSpin2k connects to this Bluetooth signal to take the cadence and power and does the rest to move the gear motor and communicate with Zwift.

This is how it works in my installation, which is very particular and does not fit in most of the bikes shown here because I had to add this KBikeBLE project into the equation.

The result I have obtained is incredible, I have been using it for more than a year and I am delighted. It works much better than commercial Smartbikes, as the Keiser M3 is a very good bike, it works spectacularly well.

In your case, you should be able to transmit this data via Bluetooth in a single signal or in separate ones with several sensors. What I don't know is how you can get SmartSpin2k to adjust the resistance of your motorized bike.

Regards.

**Reply by MortenVinding** [reply]
Yes I know, that's why I compared the KBikeBLE and the BLEBike projects :)
The BLEBike also takes the input from a reed switch on the cadence and converts it to a power rating depending on the resistance setting. Only the BLEBike does not have feedback from the bike.
My bike does however also have a potentiometer to give feedback on the resistance level (which it uses as feedback to it's own adjustments using a small motor).
Here you can see how it looks inside my bike:
![tempImageYj6ubS](https://github.com/user-attachments/assets/024d7ef7-fd5d-40da-b692-e7d71704a4dd)
![tempImage2hKSMS](https://github.com/user-attachments/assets/61718723-a9e9-4114-bb9c-87052bd3842c)
![tempImageLwhl56](https://github.com/user-attachments/assets/b837a385-2afa-4e15-8b55-a8a7d9583205)

So I guess I could use BLEBike together with SS2k more or less like you do (hacking SS2k to adjust the resistance on my bike), but I just think it's a bit clunky having to connect 3 bluetooth systems in a row :)
I would be much more elegant if SS2k could just take the cadence signal directly. After all that is pretty much what it already does with the Peloton bike (all though it's over a serial connection from what I can find out).

Ohh just found this thread about pretty much the same thing: https://github.com/doudar/SmartSpin2k/discussions/343

Btw: had a look at the M3. It certainly looks like a nice bike... but also very expensive!
I might think about upgrading my cheap secondhand bike if I get in to this Zwift thing, but right now I would like to get it working with what I have :)

Thank you so much for getting back to me 👍

**Reply by doudar** [reply]
If you add your gpio setup to main.cpp, then checkout sensorcollector.cpp and add your custom sensor to https://github.com/doudar/SmartSpin2k/tree/develop/lib/SS2K/src/sensors , (header files are in https://github.com/doudar/SmartSpin2k/tree/develop/lib/SS2K/include/sensors), you should be able to get it to work.

**Reply by MortenVinding** [reply]
> If you add your gpio setup to main.cpp, then checkout sensorcollector.cpp and add your custom sensor to https://github.com/doudar/SmartSpin2k/tree/develop/lib/SS2K/src/sensors , (header files are in https://github.com/doudar/SmartSpin2k/tree/develop/lib/SS2K/include/sensors), you should be able to get it to work.

Ohh thank a lot Anthony!
I'm unfortunately not much of a programmer, but I will defiantly give it at try.

For now I have tried to build and flash @andrewrobertdev 's fork: https://github.com/andrewrobertdev/SmartSpin2k
Fortunatly he even uses the same GPIO for cadence as me (maybe he has also tried BLEBike at some point :) )

Looks very promising but of course is quite a bit behind. But if it works it might be worth to put it up to date and make a pull request at some point?

**Comment by MortenVinding** [comment]
Just added andrewrobertdev/SmartSpin2k changes to support GPIO based cadence sensor and calculated power to my own fork:
https://github.com/MortenVinding/SmartSpin2k-gpio

This is definitely very ugly and needs to be cleaned up, both for the cadence detection, the power calculation but also to implement it in the right way... BUT it compiles which by my programming skills is a big accomplishment :)

If anyone has the time (and the bravenes) to look at it, you are more than welcome  :)

**Comment by Sc0ttmeup** [comment]
Which models of the Keiser M3 does this suit?
